pkg/auth/oauth/registry/timeoutvalidator.go

```diff
@@ -0,0 +1,267 @@
+		glog.V(5).Infof("Failed to fetch OAuthClient %q for timeout value: %v", name, err)
+		return a.defaultTimeout
+	}
+	if oauthClient.AccessTokenTimeoutSeconds == nil {
+		return a.defaultTimeout
+	}
+	a.updateTimeouts(oauthClient)
+	return timeoutAsDuration(*oauthClient.AccessTokenTimeoutSeconds)
+}
+
+func (a *TimeoutValidator) insert(td tokenData) {
+	// if there is a token we have to remove it first
+	oldtd, exists := a.data[td.token.Name]
+	if exists {
+		a.delete(oldtd, &tokenDataRef{oldtd.token.Name, oldtd.timeout()})
+	}
+	a.data[td.token.Name] = td
+	a.tree.ReplaceOrInsert(&tokenDataRef{td.token.Name, td.timeout()})
+}
+
+func (a *TimeoutValidator) delete(td tokenData, tdr *tokenDataRef) {
+	a.tree.Delete(tdr)
+	delete(a.data, td.token.Name)
+}
+
+func (a *TimeoutValidator) update(td tokenData) error {
+	delta := a.clientTimeout(td.token.ClientName)
+	newTimeout := int32(0)
+	if delta > 0 {
+		newTimeout = int32((td.seen.Sub(td.token.CreationTimestamp.Time) + delta) / time.Second)
+	}
+	// We need to get the token again here because it may have changed in the
+	// DB and we need to verify it is still worth updating
+	token, err := a.tokens.Get(td.token.Name, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if newTimeout != 0 && token.TimeoutsIn >= newTimeout {
+		// if the token was already updated with a higher or equal timeout we
+		// do not have anything to do
+		return nil
+	}
+	token.TimeoutsIn = newTimeout
+	_, err = a.tokens.Update(token)
+	return err
+}
+
+func (a *TimeoutValidator) flush(flushHorizon time.Time) {
+	flushedTokens := 0
+	totalTokens := len(a.data)
+	var retrylist []tokenData
+
+	glog.V(5).Infof("Flushing tokens timing out before %s", flushHorizon)
+
+	for item := a.tree.Min(); item != nil; item = a.tree.Min() {
+		tdr := item.(*tokenDataRef)
+		if tdr.timeout.After(flushHorizon) {
+			// out of items within the flush Horizon
+			break
+		}
+		td := a.data[tdr.name]
+		err := a.update(td)
+		switch {
+		case err == nil:
+			flushedTokens++
+		case apierrors.IsConflict(err):
+			glog.V(5).Infof("Token update deferred for token=%q, retriable error: %v",
+				td.token.Name, err)
+			retrylist = append(retrylist, td)
+		case apierrors.IsServerTimeout(err):
+			glog.V(5).Infof("Token update deferred for token=%q, retriable error: %v",
+				td.token.Name, err)
+			retrylist = append(retrylist, td)
+		default:
+			glog.V(5).Infof("Token timeout for user=%q client=%q scopes=%v was not updated: %v",
+				td.token.UserName, td.token.ClientName, td.token.Scopes, err)
+		}
+		// In all cases we remove the token from the data set
+		a.delete(td, tdr)
+	}
+
+	// we try once more and if it still fails we stop trying here and defer to a future regular update
+	for _, td := range retrylist {
+		err := a.update(td)
+		if err != nil {
+			glog.V(5).Infof("Token timeout for user=%q client=%q scopes=%v was not updated: %v",
+				td.token.UserName, td.token.ClientName, td.token.Scopes, err)
+		} else {
+			flushedTokens++
```

Comment:
Should we also count total failures?

Reply:
Any token not flushed is a failure, do we need to be very explicit about it ?